support/cas-server-support-duo-core/src/main/java/org/apereo/cas/adaptors/duo/web/flow/action/DetermineDuoUserAccountAction.java

```diff
@@ -48,11 +54,28 @@ protected Event doExecute(final RequestContext requestContext) {
             final DuoMultifactorAuthenticationProvider duoProvider = this.provider.findProvider(pr.getId(), DuoMultifactorAuthenticationProvider.class);
             final DuoSecurityAuthenticationService duoAuthenticationService = duoProvider.getDuoAuthenticationService();
             final DuoUserAccount account = duoAuthenticationService.getDuoUserAccount(p.getId());
-            if (account.getStatus() == DuoUserAccountAuthStatus.ENROLL && StringUtils.isNotBlank(duoProvider.getRegistrationUrl())) {
-                requestContext.getFlowScope().put("duoRegistrationUrl", duoProvider.getRegistrationUrl());
-                return enrollEvent;
+            switch (account.getStatus()) {
+                case ENROLL:
+                    requestContext.getFlowScope().put("duoRegistrationUrl", duoProvider.getRegistrationUrl());
+                    return enrollEvent;
+                case ALLOW:
+                    return returnByPass(authentication, duoProvider.getId());
+                case DENY:
+                    return denyEvent;
+                case UNAVAILABLE:
+                    final RegisteredServiceMultifactorPolicy.FailureModes failureMode = duoProvider.determineFailureMode(service);
+                    if (failureMode != RegisteredServiceMultifactorPolicy.FailureModes.CLOSED) {
+                        return returnByPass(authentication, duoProvider.getId());
+                    }
+                    return unavailableEvent;
             }
         }
         return success();
     }
+
+    private Event returnByPass(final Authentication authentication, final String providerId) {
+        authentication.addAttribute(AUTHENTICATION_ATTRIBUTE_BYPASS_MFA, Boolean.TRUE);
```

Comment:
I had the same thought and went looking down that path first.  
- There is no bypassProvider created currently unless one is defined in cas.properties 
- The bypassProvider is called and evaluated before MFA flow starts, before /ping or /preAuth are called.  We would need to repeat the endpoint calls again, or call bypassProvider a second time or something.
- Current bypassProvider is custom logic for specific installations, we are trying to handle bypass either from the Provider itself or because of failure mode policy.